source/server/connection_handler_impl.cc

```diff
@@ -178,17 +197,19 @@ void ConnectionHandlerImpl::ActiveSocket::continueFilterChain(bool success) {
     // Successfully ran all the accept filters.
 
     // Check if the socket may need to be redirected to another listener.
-    ActiveListener* new_listener = nullptr;
+    ActiveListenerBase* new_listener = nullptr;
 
     if (hand_off_restored_destination_connections_ && socket_->localAddressRestored()) {
       // Find a listener associated with the original destination address.
       new_listener = listener_.parent_.findActiveListenerByAddress(*socket_->localAddress());
     }
     if (new_listener != nullptr) {
+      ActiveTcpListener* tcp_listener = dynamic_cast<ActiveTcpListener*>(new_listener);
+      RELEASE_ASSERT(tcp_listener != nullptr, "ActiveSocket listner is expected to be tcp");
```

Comment:
nit: I think this can be a normal assert. It would be nice to not have a dynamic case here though. This smells to me like an interface layering issue. Maybe a TODO to see if we can do better here?

Reply:
makes sense